front/lib/resources/resource_with_vault.ts

```diff
@@ -0,0 +1,128 @@
+import type {
+  Attributes,
+  FindOptions,
+  ForeignKey,
+  Includeable,
+  ModelStatic,
+  NonAttribute,
+  WhereOptions,
+} from "sequelize";
+import { Model } from "sequelize";
+
+import type { Authenticator } from "@app/lib/auth";
+import type { Workspace } from "@app/lib/models/workspace";
+import { BaseResource } from "@app/lib/resources/base_resource";
+import type { VaultModel } from "@app/lib/resources/storage/models/vaults";
+import { VaultResource } from "@app/lib/resources/vault_resource";
+
+// Interface to enforce workspaceId and vaultId.
+interface ModelWithVault {
+  workspaceId: ForeignKey<Workspace["id"]>;
+  vaultId: ForeignKey<VaultModel["id"]>;
+  vault: NonAttribute<VaultModel>;
+}
+
+export type NonAttributeKeys<M> = {
+  [K in keyof M]: M[K] extends NonAttribute<Model<any, any>> ? K : never;
+}[keyof M] &
+  string;
+
+type InferIncludeType<M> = {
+  [K in NonAttributeKeys<M>]: M[K] extends NonAttribute<infer T>
+    ? T extends Model<any, any>
+      ? T
+      : never
+    : never;
+};
+
+export type TypedIncludeable<M> = {
+  [K in NonAttributeKeys<M>]: {
+    model: ModelStatic<InferIncludeType<M>[K]>;
+    as: K;
+  };
+}[NonAttributeKeys<M>];
+
+export interface ResourceFindOptions<M extends Model> {
+  includes?: TypedIncludeable<M>[];
+  limit?: number;
+  order?: FindOptions<M>["order"];
+  where?: WhereOptions<M>;
+}
+
+export abstract class ResourceWithVault<
+  M extends Model & ModelWithVault,
+> extends BaseResource<M> {
+  protected constructor(
+    model: ModelStatic<M>,
+    blob: Attributes<M>,
+    public readonly vault: VaultResource
+  ) {
+    super(model, blob);
+  }
+
+  protected static async baseFetch<
```

Comment:
This function is in charge of ensuring authorization given its role in I injecting the workspace id.

So let's be more explicit about it in its name. Maybe authorizedFetch rather than baseFetch?

Also since the resource that have it are in a vault we can enforce that auth has the group that owns the vault and error otherwise no?

Reply:
I initially wanted to include a permission check, but it does not work. Today, everyone needs to access this fetch those data source to list it in various places (assistant builder, Connections, ...). Additionally, system groups are never loaded in auth, so the permissions (read and write) primarily apply to writing to and reading from the data source, but does not prevent listing these items. In the future once we will have refactored everything maybe we will have a way to make this work.

Reply:
Yes that might be premature. Let's still rename the method to emphasize the fact that it plays a security role (even now with workspace filtering).

Reply:
> Yes that might be premature. Let's still rename the method to emphasize the fact that it plays a security role (even now with workspace filtering).

Isn't that similar to most resources, which are always filtered based on the workspace?

Reply:
Yes but here it's particularly important :-)

Reply:
Renamed to `baseFetchWithAuthorization`.